### PR TITLE
drivers:adc:ad7768-1: Update AD7768-1 driver

### PR DESCRIPTION
**Pull request by jgarlitos**
Update AD7768-1 Library. Added continuous read function. Change 16 bit frame to 8 bit frame. Fixed typo.

### CODE REVIEW

File: drivers/adc/ad7768-1/ad77681.c [hunk 1]
```diff
@@ -245,10 +245,10 @@ uint8_t ad77681_get_rx_buf_len(struct ad77681_dev *dev)
  * @param dev - The device structure.
  * @return frame_16bit - the number of 16 bit SPI frames
  */
-uint8_t ad77681_get_frame_16bit(struct ad77681_dev *dev)
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:248]: Can you add in the commit message why this change is needed?

File: drivers/adc/ad7768-1/ad77681.c [hunk 2]
```diff
@@ -272,39 +272,42 @@ uint8_t ad77681_get_frame_byte(struct ad77681_dev *dev)
 int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
 				  uint8_t *adc_data)
 {
-	uint8_t crc_calc_buf[4], buf[6], crc, frames_8byte = 0;
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:281]: You should have maximul of 72 characters per commit line.

File: drivers/adc/ad7768-1/ad77681.c [hunk 3]
```diff
@@ -272,39 +272,42 @@ uint8_t ad77681_get_frame_byte(struct ad77681_dev *dev)
...
+	ret = spi_write_and_read(dev->spi_desc, buf, dev->data_frame_byte + 1);
 	if (ret < 0)
 		return ret;
 
-	if (dev->crc_sel == AD77681_CRC) {
-		memcpy(crc_calc_buf, buf, ARRAY_SIZE(buf));
-		crc = ad77681_compute_crc8(crc_calc_buf, 4, INITIAL_CRC);
-		if (crc != buf[3]) {
+	if (dev->crc_sel != AD77681_NO_CRC) {
+		memcpy(crc_xor_calc_buf, buf + 1, dev->data_frame_byte - 1);
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:299]: You don't need to memcpy. Just pass buf + 1 to the compute functions

File: drivers/adc/ad7768-1/ad77681.c [hunk 4]
```diff
@@ -272,39 +272,42 @@ uint8_t ad77681_get_frame_byte(struct ad77681_dev *dev)
...
+		if (crc_xor != buf[dev->data_frame_byte]) {
 			printf("%s: CRC Error.\n", __func__);
 			ret = FAILURE;
 		}
+#ifdef CRC_DEBUG
+		char print_buf[20];
+		char ok[3] = { 'O', 'K' }, fault[6] = { 'F', 'A', 'U', 'L', 'T' };
+		sprintf(print_buf, "\n%x\t%x\tCRC/XOR: %s\n", crc_xor,
+			buf[dev->data_frame_byte],
+			((crc_xor != buf[dev->data_frame_byte]) ? (fault) : (ok)));
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:316]: ((crc_xor != buf[dev->data_frame_byte]) ? "FAULT" : "OK"));

Comment (mchindri) [reply drivers/adc/ad7768-1/ad77681.c:316]: and you can printf directly without using printf_buf

File: drivers/adc/ad7768-1/ad77681.c [hunk 5]
```diff
@@ -324,6 +324,57 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
 	return ret;
 }
 
+/**
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:327]: Max 70 chars in commit messaje per line

File: drivers/adc/ad7768-1/ad77681.c [hunk 6]
```diff
@@ -324,6 +324,57 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+ * Read continuous conversion result from device.
+ * @param dev - The device structure.
+ * @param adc_data - The conversion result data
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_spi_read_adc_cont_data(struct ad77681_dev *dev,
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:333]: This functions is very similar to ad77681_spi_read_adc_data.
Can you create just one function and control with one parameter if it is continuous read?
This function can be internal to the driver, an both ad77681_spi_read_adc_data and ad77681_spi_read_adc_cont_data call it

File: drivers/adc/ad7768-1/ad77681.c [hunk 7]
```diff
@@ -272,39 +272,42 @@ uint8_t ad77681_get_frame_byte(struct ad77681_dev *dev)
...
-	if (dev->status_bit)
-		frames_8byte++;
-
 	buf[0] = AD77681_REG_READ(AD77681_REG_ADC_DATA);
 	buf[1] = 0x00;
 	buf[2] = 0x00;
 	buf[3] = 0x00;
 	buf[4] = 0x00;	/* added 2 more array places for max data length read */
-	buf[5] = 0x00;	/* register address + 3 bytes of data (24bit format) + CRC + Status bit */
+	buf[5] = 0x00;	/* register address + 3 bytes of data (24bit format) + Status bit + CRC */
```
Comment (adrimbarean) [inline drivers/adc/ad7768-1/ad77681.c:291]: nitpick: addition operands are interchangeable so technically no need for this fix.

File: drivers/adc/ad7768-1/ad77681.c [hunk 8]
```diff
@@ -278,39 +280,42 @@ uint8_t ad77681_get_frame_16bit(struct ad77681_dev *dev)
...
-	if (dev->crc_sel == AD77681_CRC) {
-		memcpy(crc_calc_buf, buf, ARRAY_SIZE(buf));
-		crc = ad77681_compute_crc8(crc_calc_buf, 4, INITIAL_CRC);
-		if (crc != buf[3]) {
+	if (dev->crc_sel != AD77681_NO_CRC) {
+		memcpy(crc_xor_calc_buf, buf + 1, dev->data_frame_byte - 1);
+		if (dev->crc_sel == AD77681_CRC)
+			crc_xor = ad77681_compute_crc8(crc_xor_calc_buf, dev->data_frame_byte - 1,
+						       INITIAL_CRC_CRC8);
+		else if (dev->crc_sel == AD77681_XOR)
```
Comment (adrimbarean) [inline drivers/adc/ad7768-1/ad77681.c:303]: It is a good practice to end every else if chain with a final else to cover all cases. In this case it will help with the cpp_check error too.

File: drivers/adc/ad7768-1/ad77681.c [hunk 9]
```diff
@@ -319,6 +324,57 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	ret = spi_write_and_read(dev->spi_desc, buf, dev->data_frame_byte);
+	if (ret < 0)
+		return ret;
+
+	if (dev->crc_sel != AD77681_NO_CRC) {
+		memcpy(crc_xor_calc_buf, buf, dev->data_frame_byte - 1);
+		if (dev->crc_sel == AD77681_CRC)
+			crc_xor = ad77681_compute_crc8(crc_xor_calc_buf, dev->data_frame_byte - 1,
+						       INITIAL_CRC_CRC8);
+		else if (dev->crc_sel == AD77681_XOR)
```
Comment (adrimbarean) [inline drivers/adc/ad7768-1/ad77681.c:354]: 'else if' here too.

File: drivers/adc/ad7768-1/ad77681.c (hunk 1 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c:248]: Done adding comments

File: drivers/adc/ad7768-1/ad77681.c (hunk 2 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c:281]: Done with the edits of the commit message.

File: drivers/adc/ad7768-1/ad77681.c (hunk 3 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edits.

File: drivers/adc/ad7768-1/ad77681.c (hunk 4 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edits.

File: drivers/adc/ad7768-1/ad77681.c (hunk 5 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c:319]: Done with the edits of the commit message.

File: drivers/adc/ad7768-1/ad77681.c (hunk 6 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c:325]: I think it would be better to have a separate "ad77681_spi_read_adc_cont_data " function. This function will be called thousand times in a seconds when reading the continuously streamed data from the ADC. Merging the functions, the controller would also check thousand of times in a sec the type of read (register or continuous read) will be performed if operating on a continuous read mode.

File: drivers/adc/ad7768-1/ad77681.c (hunk 7 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edits.

File: drivers/adc/ad7768-1/ad77681.c (hunk 8 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edit.

File: drivers/adc/ad7768-1/ad77681.c (hunk 9 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edit.

**Review by amiclaus — APPROVED** [review]

File: drivers/adc/ad7768-1/ad77681.c [hunk 10]
```diff
@@ -272,20 +272,9 @@ uint8_t ad77681_get_frame_byte(struct ad77681_dev *dev)
 int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
 				  uint8_t *adc_data)
 {
-	uint8_t crc_calc_buf[4], buf[6], crc, frames_8byte = 0;
+	uint8_t crc_xor_calc_buf[4], buf[6], crc_xor;
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:279]: crc_xor_calc_buf is not needed anymore

File: drivers/adc/ad7768-1/ad77681.c [hunk 11]
```diff
@@ -316,6 +316,54 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	buf[2] = 0x00;
+	buf[3] = 0x00;	/* added 2 more array places for max data length read */
+	buf[4] = 0x00;	/* 3 bytes of data (24bit format) + Status bit + CRC */
+
+	ret = spi_write_and_read(dev->spi_desc, buf, dev->data_frame_byte);
+	if (ret < 0)
+		return ret;
+
+	if (dev->crc_sel != AD77681_NO_CRC) {
+		memcpy(crc_xor_calc_buf, buf, dev->data_frame_byte - 1);
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:342]: Same here: https://github.com/analogdevicesinc/no-OS/pull/957#discussion_r588077589

File: drivers/adc/ad7768-1/ad77681.c (hunk 6 above)
Comment (mchindri) [reply drivers/adc/ad7768-1/ad77681.c:325]: I see, I think from the maintenance point of view will be good to don't have duplicated code. A good example is https://github.com/analogdevicesinc/no-OS/pull/957/commits/7bd24bfc17d3a92128f3073fb29830e46c7e2f43#r589349134
.
Anyway maybe someone else of @analogdevicesinc/noos team have other opinion.

File: drivers/adc/ad7768-1/ad77681.c (hunk 10 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edits.

File: drivers/adc/ad7768-1/ad77681.c (hunk 11 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done with the edits.

File: drivers/adc/ad7768-1/ad77681.c [hunk 12]
```diff
@@ -316,6 +316,53 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	buf[3] = 0x00;	/* added 2 more array places for max data length read */
+	buf[4] = 0x00;	/* 3 bytes of data (24bit format) + Status bit + CRC */
+
+	ret = spi_write_and_read(dev->spi_desc, buf, dev->data_frame_byte);
+	if (ret < 0)
+		return ret;
+
+	if (dev->crc_sel != AD77681_NO_CRC) {
+		if (dev->crc_sel == AD77681_CRC)
+			crc_xor = ad77681_compute_crc8(buf, dev->data_frame_byte - 1,
```
Comment (mchindri) [inline drivers/adc/ad7768-1/ad77681.c:343]: In the ad77681_spi_read_adc_data was buf + 1 shouldn't be here the same? or there just buf?

Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c:343]: It is only buf. The ADC data will be stored starting buf[0] since there will be no more register writes. Just flushing out of date using spi sclk. CRC is also stored at the last buf element of spi transaction (buf[dev->data_frame_byte-1]).

File: drivers/adc/ad7768-1/ad77681.c (hunk 6 above)
Comment (adrimbarean) [reply drivers/adc/ad7768-1/ad77681.c:325]: I will agree with Mihail here. The function could be `inline` if call times are an issue.

**Review by adrimbarean — APPROVED** [review]
Looks good to me. Just one comment, but will approve to not hinder the progress if we decide that.

File: drivers/adc/ad7768-1/ad77681.c (hunk 6 above)
Comment (jgarlitos) [reply drivers/adc/ad7768-1/ad77681.c]: Done merging two functions. Please check if the implementation is okay.

**Review by dbogdan — APPROVED** [review]
